Tests to validate resources deployed by etcd-druid

georgekuruvillak opened the pull request:
**What this PR does / why we need it**:
This PR consists of test to validate the functionality of tests.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```improvement operator
NONE
```

amshuman-kr commented on controllers/etcd_controller_test.go:
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
+	Expect(err).NotTo(HaveOccurred())
+	images, err := imagevector.FindImages(imageVector, imageNames)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(containers[common.Etcd].Image).To(Equal(fmt.Sprintf("%s:%s", images[common.Etcd].Repository, *images[common.Etcd].Tag)))
```
https://onsi.github.io/gomega/#consistofelement-interface?
https://onsi.github.io/gomega/#containelementelement-interface?

georgekuruvillak replied:
It is just a string comparison here. Equal should suffice, ryt?

amshuman-kr commented on controllers/etcd_controller_test.go:
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
+
+	// Validate Metrics MetricsLevel
+	Expect(instance.Spec.Etcd.Metrics).To(BeNil())
+	val, ok = config[metricsKey]
```
https://onsi.github.io/gomega/#havekeywithvaluekey-interface-value-interface

georgekuruvillak commented on controllers/etcd_controller_test.go:
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
 
 }
 
+func validateEtcdConfigWithDefaults(s *appsv1.StatefulSet, cm *corev1.ConfigMap, svc *corev1.Service, instance *druidv1alpha1.Etcd) {
```
I have removed all the loops and conditionals and replaced them with gomega structs. The tests are much cleaner now. PTAL

georgekuruvillak commented on controllers/etcd_controller_test.go:
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
+	Expect(len(livelinessProbe.Exec.Command)).To(Equal(7))
+}
+
+func validateEtcdConfig(s *appsv1.StatefulSet, cm *corev1.ConfigMap, svc *corev1.Service, instance *druidv1alpha1.Etcd) {
```
I have removed all the loops and conditionals and replaced them with gomega structs. The tests are much cleaner now. PTAL

georgekuruvillak commented:
@AndreasBurger Can you check out the pipeline_definition file in `.ci` directory.

amshuman-kr commented:
@georgekuruvillak I don't see any changes related to gomega structs.

georgekuruvillak commented:
@amshuman-kr I haven't added gstruct. Instead, I have removed all the loops and conditionals and replaced them with gomega structs. The tests look much cleaner now. I shall add in gostruct too.

amshuman-kr commented:
@georgekuruvillak I didn't understand what you meant by gomega structs then? I still see loops and conditions.

georgekuruvillak commented:
The `if` condition is used [here]( https://github.com/gardener/etcd-druid/pull/41/files#diff-03838f4ddc16876e7feda81a4bedf7d8R467) as the `tls` fields are used to check the liveliness probes and readiness probes. The `for` loop is used [here](https://github.com/gardener/etcd-druid/pull/41/files#diff-03838f4ddc16876e7feda81a4bedf7d8R427) for choosing the container to test against. By go omega struct I meant the constructs in gomega like `ConsistOf`, `HasKeyWithValue`. I havent implemented gstruct.

swapnilgm commented:
@georgekuruvillak Please remove pipeline definitions updates. Those are fixed with https://github.com/gardener/etcd-druid/pull/58

georgekuruvillak commented:
@amshuman-kr I have made the changes. PTAL.

georgekuruvillak commented on controllers/constants_test.go:
```diff
@@ -0,0 +1,179 @@
+var (
+	bootstrapScriptWithoutTLS = scriptWithoutTLS + startManagedEtcd + checkAndStartEtcdWithoutTLS + bootstrapWithValidation
+	bootstrapScriptWithTLS    = scriptWithTLS + startManagedEtcd + checkAndStartEtcdWithTLS + bootstrapWithValidation
+	cmdIterator               = func(element interface{}) string {
```
Makes sense. Ill make sthe change and update the PR.

swapnilgm commented:
@georgekuruvillak Will you please squash the commits.
